fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/MetricUnionAvailability.java

```diff
@@ -0,0 +1,190 @@
+                );
+    }
+
+    @Override
+    public SimplifiedIntervalList getAvailableIntervals(DataSourceConstraint constraints) {
+        return new SimplifiedIntervalList(
+                constructSubConstraint(constraints).entrySet().stream()
+                        .map(entry -> entry.getKey().getAvailableIntervals(entry.getValue()))
+                        .map(i -> (Set<Interval>) new HashSet<>(i))
+                        .reduce(null, IntervalUtils::getOverlappingSubintervals)
+        );
+    }
+
+    /**
+     * Returns duplicate values of <tt>MetricColumn</tt>'s in a map of <tt>Availability</tt>
+     * to a set of <tt>MetricColumn</tt>'s contained in that <tt>Availability</tt>.
+     * <p>
+     * The return value is a map of duplicate <tt>MetricColumn</tt> to all <tt>Availability</tt>'s that contains
+     * this <tt>MetricColumn</tt>
+     *
+     * @param availabilitySetMap  A map from <tt>Availability</tt> to set of <tt>MetricColumn</tt> contained in that
+     * <tt>Availability</tt>
+     *
+     * @return duplicate values of <tt>MetricColumn</tt>'s
+     */
+    private static Map<MetricColumn, Set<Availability>> getDuplicateValue(
+            Map<Availability, Set<MetricColumn>> availabilitySetMap
+    ) {
+        Map<MetricColumn, Set<Availability>> metricColumnSetMap = new HashMap<>();
+
+        for (Map.Entry<Availability, Set<MetricColumn>> entry : availabilitySetMap.entrySet()) {
+            Availability availability = entry.getKey();
+            for (MetricColumn metricColumn : entry.getValue()) {
+                metricColumnSetMap.put(
+                        metricColumn,
+                        Sets.union(
+                                metricColumnSetMap.getOrDefault(metricColumn, new HashSet<>()),
+                                Sets.newHashSet(availability)
+                        )
+                );
+            }
+        }
+
+        return metricColumnSetMap.entrySet().stream()
+                .filter(entry -> entry.getValue().size() > 1)
+                .collect(
+                        Collectors.toMap(
+                                Map.Entry::getKey,
+                                Map.Entry::getValue
+                        )
+                );
+    }
+
+    /**
+     * Construct a mapping from a type of <tt>Availability</tt> to <tt>DataSourceConstraint</tt> of that availability
+     * and filter out mappings that maps to <tt>DataSourceConstraint</tt> with empty set of metric names.
+     *
+     * @param constraint data constraint containing columns and api filters
+     *
+     * @return A mapping from <tt>Availability</tt> to <tt>DataSourceConstraint</tt> with non-empty metric names
+     */
+    private Map<Availability, DataSourceConstraint> constructSubConstraint(DataSourceConstraint constraint) {
+        return getAvailabilityToConstraintMapping(constraint).entrySet().stream()
+                .filter(entry -> !entry.getValue().getMetricNames().isEmpty())
```

Comment:
We seem to be filtering these maps based on values (or keys) a whole lot. Perhaps we can combine some of these streams into a single stream, rather than going to an intermediate collection every time?